### PR TITLE
Add capstyle and joinstyle attributes to Collection class (Issue #8277)

### PR DESCRIPTION
**Pull request by simonpf**
## PR Summary

Adds  `_capstyle` and `_joinstyle` attributes as well as corresponding setters to `Collection` class. Changes  `draw` class method to set capstyle and joinstyle of renderer accordingly. This solves the problem of not being able to set capstyle and joinstyle of `LineCollection` ([Issue #8277](https://github.com/matplotlib/matplotlib/issues/8277)).

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

## Example

```
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection
from matplotlib import colors as mcolors

import numpy as np

x = np.arange(100)
ys = x[::10, np.newaxis] + (x + 10.0 * np.random.rand(*x.shape))[np.newaxis, :]

segs = np.zeros((10, 100, 2), float)
segs[:, :, 1] = ys
segs[:, :, 0] = x

segs = np.ma.masked_where((segs > 50) & (segs < 60), segs)
ax = plt.axes()
ax.set_xlim(x.min(), x.max())
ax.set_ylim(ys.min(), ys.max())

colors = [mcolors.to_rgba(c)
          for c in plt.rcParams['axes.prop_cycle'].by_key()['color']]
line_segments = LineCollection(segs, linewidths=(2, 3, 4, 5),
                               colors=colors, linestyle='solid')
line_segments.set_joinstyle("miter")
line_segments.set_capstyle("round")
ax.add_collection(line_segments)
ax.set_title('Line collection with masked arrays')
plt.show()
```

### CODE REVIEW

**Comment by anntzer** [comment]
Unless I am mistaken, this does not actually cause the capstyle and joinstyle to be used while drawing.  You probably need to update Collection.draw accordingly.

Note that unless you mess with Renderer.draw_path_collection, the capstyle and joinstyle will be global ones (not settable per-item).  On the one hand, who wants to set these per item?  On the other hand, even antialiasing is settable per-item.

**Comment by simonpf** [comment]
Hi @anntzer,

I did update change `Collection.draw` and tested this with the example given in the issue and the one above.

Yes, I was also wondering if it should be possible to set cap- and joinstyle for each element in the collection, but as I understand it, this is not the solution proposed in the issue.

**Comment by anntzer** [comment]
Sorry, I was not careful when proofreading.  Indeed they are set correctly.

Allowing per-item setting is a much more complicated piece of work, so I think I'm OK with not having it for now.

I think this warrants an image test (to show that the styles are actually propagated to the draw), and a whatsnew entry.

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -104,6 +104,8 @@ def __init__(self,
                  facecolors=None,
                  linewidths=None,
                  linestyles='solid',
+                 capstyle=None,
```
Comment (NelleV) [inline lib/matplotlib/collections.py]: Those two new arguments need to be documented in the docstring for this class.

File: lib/matplotlib/collections.py [hunk 2]
```diff
@@ -536,6 +554,42 @@ def set_linestyle(self, ls):
         self._linewidths, self._linestyles = self._bcast_lwls(
             self._us_lw, self._us_linestyles)
 
+    def set_capstyle(self, cs):
+        """
+        Set the capstyle for the collection.
+
+        ACCEPTS: ['butt' | 'round' | 'projecting']
```
Comment (NelleV) [inline lib/matplotlib/collections.py]: We're using numpydoc style docstring. Can you delete this line, but add this to the parameters list? (See comment bellow).

File: lib/matplotlib/collections.py [hunk 3]
```diff
@@ -536,6 +554,42 @@ def set_linestyle(self, ls):
...
+
+        Parameters
+        ----------
+        cs : The capstyle
```
Comment (NelleV) [inline lib/matplotlib/collections.py]: Considering we are using numpydoc, that should be:
```
cs : ['butt' | 'round' | 'projecting']
  the capstyle
```

File: lib/matplotlib/collections.py [hunk 4]
```diff
@@ -536,6 +554,42 @@ def set_linestyle(self, ls):
...
+            raise ValueError('Unrecognized cap style.  Found %s' % cs)
+
+    def get_capstyle(self):
+        return self._capstyle
+
+    def set_joinstyle(self, js):
+        """
+        Set the joinstyle for the collection.
+
+        ACCEPTS: ['miter' | 'round' | 'bevel']
```
Comment (NelleV) [inline lib/matplotlib/collections.py]: Same here :)

File: lib/matplotlib/collections.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/collections.py]: Until we *actually* get rid of the kwdoc mechanism (whether we do so is a separate issue), I would prefer leaving the ACCEPTS strings there, as it avoids having an ugly `property -- unknown` entry in the rendered table.  (But that does not preclude documenting them in the parameters table below *too*.)

File: lib/matplotlib/collections.py [hunk 5]
```diff
@@ -52,11 +52,16 @@ class Collection(artist.Artist, cm.ScalarMappable):
...
+    Exceptions are *capstyle* and *joinstyle* properties, these can
+    only be set globally for the whole collection.
+
     Keyword arguments and default values:
 
         * *edgecolors*: None
         * *facecolors*: None
         * *linewidths*: None
+        * *capstyle*:   None
+        * *joinstyle*:  None
```
Comment (NelleV) [inline lib/matplotlib/collections.py]: :+1:

**Comment by simonpf** [comment]
Sorry for the noise, I had some problems with my test setup and missed the PEP8 warnings.

**Comment by simonpf** [comment]
So is there anything left to do on this?

File: doc/users/whats_new.rst [hunk 6]
```diff
@@ -246,6 +246,13 @@ volumetric model.
 Improvements
 ++++++++++++
 
+Add `capstyle` and `joinstyle` attributes to `Collection`
```
Comment (anntzer) [inline doc/users/whats_new.rst]: capstyle and joinstyle should use *two* backquotes on both sides (this is somewhat poorly documented at http://matplotlib.org/devdocs/devel/documenting_mpl.html#formatting but the idea is that single backquotes mean "a python object that has its own entry in the docs", which is the case for Collection but not for capstyle and joinstyle, which are just parameters).

Same below.

File: lib/matplotlib/tests/test_collections.py [hunk 7]
```diff
@@ -626,6 +626,64 @@ def test_lslw_bcast():
...
+def test_capstyle():
+    col = mcollections.PathCollection([], capstyle='round')
+    assert_equal(col.get_capstyle(), 'round')
+    col.set_capstyle('butt')
+    assert_equal(col.get_capstyle(), 'butt')
+
+
+@image_comparison(baseline_images=['capstyle'],
+                  extensions=['png'])
+def test_capstyle_image():
```
Comment (anntzer) [inline lib/matplotlib/tests/test_collections.py]: Can you join the two image comparisons into a single test and thus a single reference image?  Reference test images are a large part of why the repo is so heavy, so we're trying to be careful with adding more of them.
After doing so, you can squash the old images out of the history with rebase; if you're not comfortable with it I can take care of it when merging too, just let me know.

**Comment by anntzer** [comment]
Just some minor stuff.  The rest looks good.

**Comment by simonpf** [comment]
OK, fixed now.

**Comment by NelleV** [comment]
Thanks for the contribution @simonpf !

File: doc/users/whats_new.rst [hunk 8]
```diff
@@ -246,6 +246,13 @@ volumetric model.
 Improvements
 ++++++++++++
 
+Add ``capstyle`` and ``joinstyle`` attributes to `Collection`
```
Comment (tacaswell) [inline doc/users/whats_new.rst]: This got put into the 2.1 whats_new :disappointed:

**Pull request merged**